### ISSUE TITLE
[iOS] launchImageLibrary subsequent calls do not work after being called once

### ISSUE THREAD

**Issue by scousino** [react-native-image-picker/react-native-image-picker#1456]
### Description
On iOS, I can make a call to launchImageLibrary, select an image/video, and then do some stuff with it. Now I want to call launchImageLibrary and select a new image/video, however after the second call to launchImageLibrary nothing appears on the screen and the callback passed is never called with any error

### How to repeat issue and example
* call launchImageLibrary
* either cancel or choose a picture
* call launchImageLibrary again with the exact same arguments
* nothing happens and no error is given

### Solution

Subsequent calls to launchImageLibrary should work, regardless of whether its been called already

### Additional Information

* Image Picker version: v3.0.0-vnext.1
* React Native version: 0.62.2
* Platform: iOS
* Development Operating System: MacOS
* Dev tools: XCode 12
* iOS version: 13.4.1

**Comment by scousino** [comment]
This is also an issue on the stable release version of 2.3.1

**Comment by scousino** [comment]
I've found some additional logs that help track down the issue. It seems a lot of calls on the ios side related to the UI are not happening on the main thread. These logs appear immediately after the first call to launchImageLibrary, but do not appear on subsequent calls to the same method.

[Logs.txt](https://github.com/react-native-image-picker/react-native-image-picker/files/5460549/Logs.txt)

**Comment by ravirajn22** [comment]
I am not able to reproduce your issue on vnext branch. Show your code to debug futher.

**Comment by cpecorari** [comment]
Hi, just upgraded from 2.3.1 to vnext.2, and so far so good, looks promising !
No issue as @scousino mentioned...
Would you share some code to help you ?

**Comment by scousino** [comment]
Hi, there isn't much code to show as its a simple interaction. The only other possible thing I can think of is I show a modal selector from react-native-modal-selector, and then open the camera or the image library, based on the selection. However, like I mentioned it always opens on the first time, but never after that. On Android I do not see this specific problem.

This is how I launch the image library.
```
launchImageLibrary(
                { mediaType: mediaType.imagePickerMediaType, quality: 1 },
                (res) => imgPickerCallback(res, mediaType.enumMediaType, true),
            );
```

**Comment by ravirajn22** [comment]
There must be some conflict between react-native-modal-selector and iOS picker. Make sure to close the react-native-modal-selector and then call launchImageLibrary probably for now use  `setTimeout` to wait for the modal to close or we have callback like onModalClose etc then call this library on that.

**Comment by scousino** [comment]
Yeah I thought the same thing and already have that implemented. That is how I got it to open the image library at all. Now, with the setTimeout, I'm running into my described issue. My callback, imPickerCallback isn't even called with an error on subsequent launchImageLibrary calls. So, something in that call bails out, but with no error.

**Comment by ravirajn22** [comment]
I may not be able to help you, I have tested this on react-native modal it works. I think your issue is out of scope of this library.

**Comment by scousino** [comment]
I have found a new log that I think should shed light on the issue: 
Attempt to present <UIImagePickerController: 0x10b03dc00> on <RCTModalHostViewController: 0x113f8e0f0> (from <RCTModalHostViewController: 0x113f8e0f0>) whose view is not in the window hierarchy.

**Comment by scousino** [comment]
also @ravirajn22 I used react-native-modal-selector, not react-native-modal

**Comment by Johan-dutoit** [comment]
@scousino might be worth trying to reproduce in the example app and also comparing differences.

**Comment by scousino** [comment]
I'll try this now, as a side note I would recommend adding a README about how to run the example app

**Comment by scousino** [comment]
@Johan-dutoit and @ravirajn22 I've tried to reproduce on the example app, and the side effect of calling launchImageLibrary after a modal selection, while using a setTimeout as recommended above, still shows incorrect behavior. Now, the example app just crashes out entirely. 

I've created a fork and a branch on that fork with my changes [here](https://github.com/scousino/react-native-image-picker/tree/scousino-1456)

**Comment by Johan-dutoit** [comment]
> Also I'm not sure the example app should have code that references the local react-native-image-picker package. Wouldn't it make more sense to use released and deployed versions via npm?

I'm in agreement with you, It's just been easier to test local changes while it still points locally.  I think once we've reached a stable release, then it's definitely worth changing over to the npm version.

**Comment by marcoautiero** [comment]
> I have found a new log that I think should shed light on the issue:
> Attempt to present <UIImagePickerController: 0x10b03dc00> on <RCTModalHostViewController: 0x113f8e0f0> (from <RCTModalHostViewController: 0x113f8e0f0>) whose view is not in the window hierarchy.

Same problem here, this module is not usable in this condition...

**Comment by ravirajn22** [comment]
When using react-native modal or derivate libraries to show the options and closing them, Make sure the Modal is closed completely before calling the library methods. Because the library opens a View which attaches to previous view, since the previous view is closed the `launchImageLibrary` or `launchCamera` cannot attach to any view and the error happens. This is a known in issue in react-native when dealing with Closing and opening a iOS modal simultaneously.

**Comment by trongtai1801** [comment]
in my case, I set timeout and it work well
<img width="320" alt="Screen Shot 2020-12-13 at 1 31 55 AM" src="https://user-images.githubusercontent.com/31733095/101992029-fe339180-3ce2-11eb-8fba-376bfe72e5ee.png">

**Comment by nkpatnis** [comment]
This solved my issue
```
  launchSelection(func) {
    this.setState({ modalVisible: false }, () => setTimeout(func, 1000));
  }
```

**Comment by dpgailey** [comment]
This isn't really a work around. Adding a timeout negatively impacts UX. Is there a correct work around or perhaps a fix?

**Comment by gburnett** [comment]
> When using react-native modal or derivate libraries to show the options and closing them, Make sure the Modal is closed completely before calling the library methods. Because the library opens a View which attaches to previous view, since the previous view is closed the `launchImageLibrary` or `launchCamera` cannot attach to any view and the error happens. This is a known in issue in react-native when dealing with Closing and opening a iOS modal simultaneously.

@ravirajn22 How do you make sure that the Modal is closed?

**Comment by dika4152** [comment]
> This solved my issue
> 
> ```
>   launchSelection(func) {
>     this.setState({ modalVisible: false }, () => setTimeout(func, 1000));
>   }
> ```

this one is works for me. i also change the timeout duration to 300 for better UX and it works well too.

**Comment by bibin96** [comment]
> using timeout when calling the launchCamera or launchImageLibrary after update modal state to false is work for me. and I set timeout duration to 500 for better UX and it works well too
> 
> ```
> () => {
>   await setShowOption(false);
>   setTimeout(() => launchCamera(cameraOption, responseHandler), 500)
> }
> ```

This worked for me

**Comment by scousino** [comment]
> 
> 
> This isn't really a work around. Adding a timeout negatively impacts UX. Is there a correct work around or perhaps a fix?
> 
> Is there a way to build in callbacks for when modals are effectively off-screen and cleaned up?

I believe this should be the actual solution that is looked into or something of that nature. Using timeouts is a hack that creates dumb UX

**Comment by quiringk** [comment]
This was the solution for me today. Everything seemed to be working, I would select the picture, the picture would load in my UI and then the app would freeze and I couldn't click anything. This was very hard to debug.
